File: 3583-sorted-gcd-pair-queries/sorted-gcd-pair-queries.py
import math
import bisect

class Solution:
    def gcdValues(self, nums: list[int], queries: list[int]) -> list[int]:
        # Track the largest element to know the absolute upper boundary for our sieve calculations
        max_val = max(nums)
        
        # Build a frequency map to record how many times each number appears in the dataset
        counts = [0] * (max_val + 1)
        for x in nums:
            counts[x] += 1
            
        # Set up an array to calculate the exact frequency of pairs that produce a specific GCD value
        exact_gcd_pairs = [0] * (max_val + 1)
        
        # Iterate backwards to filter out contributions from higher multiples using inclusion-exclusion logic
        for i in range(max_val, 0, -1):
            multiples_count = 0
            
            # Sum up frequencies of all numbers in the dataset that are perfectly divisible by 'i'
            for j in range(i, max_val + 1, i):
                multiples_count += counts[j]
                
            # Compute total theoretical pair combinations that share 'i' as a common divisor
            total_pairs_with_divisor = (multiples_count * (multiples_count - 1)) // 2
            
            # Subtract counts belonging to larger multiples of 'i' to isolate pairs where GCD is exactly 'i'
            for j in range(2 * i, max_val + 1, i):
                total_pairs_with_divisor -= exact_gcd_pairs[j]
                
            exact_gcd_pairs[i] = total_pairs_with_divisor
            
        # Accumulate exact pair counts into a running prefix sum array to mark index distribution boundaries
        prefix_sum = [0] * (max_val + 1)
        for i in range(1, max_val + 1):
            prefix_sum[i] = prefix_sum[i - 1] + exact_gcd_pairs[i]
            
        # Map each target query position instantly to its corresponding GCD value using binary searching
        return [bisect.bisect_right(prefix_sum, q) for q in queries]

File: 3583-sorted-gcd-pair-queries/sorted-gcd-pair-queries.c
#include <stdlib.h>

/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* gcdValues(int* nums, int numsSize, long long* queries, int queriesSize, int* returnSize) {
    // Determine the maximum element in the dataset to set limits for our sieve array sizes
    int maxVal = 0;
    for (int i = 0; i < numsSize; i++) {
        if (nums[i] > maxVal) {
            maxVal = nums[i];
        }
    }

    // Allocate an array to count the occurrence frequency of each number up to maxVal
    int* counts = (int*)calloc(maxVal + 1, sizeof(int));
    for (int i = 0; i < numsSize; i++) {
        counts[nums[i]]++;
    }

    // Allocate space to store the exact number of pairs yielding a specific GCD value
    long long* exactGcdPairs = (long long*)calloc(maxVal + 1, sizeof(long long));
    
    // Process backwards to easily filter out contributions belonging to higher multiples
    for (int i = maxVal; i >= 1; i--) {
        long long multiplesCount = 0;
        
        // Sum up frequencies of all numbers that are perfectly divisible by 'i'
        for (int j = i; j <= maxVal; j += i) {
            multiplesCount += counts[j];
        }
        
        // Compute total combination choices of pairs that share 'i' as a common divisor
        long long totalPairsWithDivisor = (multiplesCount * (multiplesCount - 1)) / 2;
        
        // Subtract counts belonging to larger multiples of 'i' to isolate pairs where GCD is exactly 'i'
        for (int j = 2 * i; j <= maxVal; j += i) {
            totalPairsWithDivisor -= exactGcdPairs[j];
        }
        exactGcdPairs[i] = totalPairsWithDivisor;
    }

    // Build a prefix sum array to keep track of the cumulative number of pairs up to GCD 'i'
    long long* prefixSum = (long long*)calloc(maxVal + 1, sizeof(long long));
    for (int i = 1; i <= maxVal; i++) {
        prefixSum[i] = prefixSum[i - 1] + exactGcdPairs[i];
    }

    // Allocate memory for the final output array that will hold the answers for each query index
    int* answer = (int*)malloc(queriesSize * sizeof(int));
    *returnSize = queriesSize; // Tell the caller the absolute size of the returned array

    // Process each query position via a custom upper_bound binary search over our distribution table
    for (int i = 0; i < queriesSize; i++) {
        long long targetIndex = queries[i];
        
        int low = 1, high = maxVal, result = maxVal;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            
            // If prefix sum is strictly greater, it covers the index. Record it and try to search left.
            if (prefixSum[mid] > targetIndex) {
                result = mid;
                high = mid - 1; 
            } else {
                // Otherwise, the current prefix window isn't big enough yet. Search right.
                low = mid + 1;
            }
        }
        answer[i] = result;
    }

    // Free local dynamic storage buffers to prevent memory leaks before returning the response
    free(counts);
    free(exactGcdPairs);
    free(prefixSum);

    return answer;
}